Source/Dafny/ConcreteSyntax/ConcreteSyntaxTree.cs

```diff
@@ -14,65 +19,90 @@ public class ConcreteSyntaxTree : ICanRender {
 
     private readonly IList<ICanRender> _nodes = new List<ICanRender>();
 
+    public IEnumerable<ICanRender> Nodes => _nodes;
+
     public ConcreteSyntaxTree Fork(int relativeIndent = 0)
     {
       var result = new ConcreteSyntaxTree(relativeIndent);
       _nodes.Add(result);
       return result;
     }
 
+    public T Prepend<T>(T node) 
+      where T : ICanRender {
+      _nodes.Insert(0, node);
+      return node;
+    }
+    
     public T Append<T>(T node) 
       where T : ICanRender {
       Contract.Requires(node != null);
       _nodes.Add(node);
       return node;
     }
-
-    public void Write(object value) {
-      Write(value.ToString());
-    }
         
-    public void Write(string value) {
+    public ConcreteSyntaxTree Write(string value) {
       _nodes.Add(new LineSegment(value));
+      return this;
     }
 
-    public void WriteLine(string format, params object[] args) {
-      Write(format, args);
-      WriteLine();
+    [StringFormatMethod("format")]
+    public ConcreteSyntaxTree WriteLine(string format, params object[] args) {
+      WriteLine(string.Format(format, args));
+      return this;
     }
-    
-    public void WriteLine(string value) {
+
+    public ConcreteSyntaxTree WriteLine(string value) {
       Write(value);
       WriteLine();
+      return this;
     }
         
-    public void WriteLine() {
+    public ConcreteSyntaxTree WriteLine() {
       _nodes.Add(new NewLine());
+      return this;
     }
 
-    public void Write(string format, params object[] args) {
+    [StringFormatMethod("format")]
+    public ConcreteSyntaxTree Write(string format, params object[] args) {
       Write(string.Format(format, args));
+      return this;
+    }
+
+    public ConcreteSyntaxTree FormatLine(FormattableString input) {
+      Format(input);
+      return WriteLine();
+    }
+
+    public ConcreteSyntaxTree Format(FormattableString input) {
+      var anchorString = string.Format(input.Format, Enumerable.Range(0, input.ArgumentCount).
+        Select(index => $"magicAnchor${index}").ToArray<object>());
+      for (int argIndex = 0; argIndex < input.ArgumentCount; argIndex++) {
+        var split = anchorString.Split($"magicAnchor${argIndex}");
+        anchorString = split.Length > 0 ? split[1] : "";
+        Write(split[0]);
+        object argument = input.GetArgument(argIndex)!;
+        if (argument is ConcreteSyntaxTree treeArg) {
+          Append(treeArg);
+        } else {
+          Write(argument.ToString());
+        }
+      }
```

Comment:
I'm not sure what to add in a comment apart from the method signature. It's like `Write` and `Append` but takes a `FormattableString`. I would have called it `Write` but that caused issues with overloading since `FormattableString` can be implicitly cast to a `String`.